### PR TITLE
Added List keyboard to Calculator dialog

### LINKED ISSUE #7886: Adding a new lists keyboard into the calculator
Here is an example of a dataset, where a variable contains a set of simple expressions
[superabundant20.xlsx](https://github.com/africanmathsinitiative/R-Instat/files/9597668/superabundant20.xlsx)
Here it is in R-Instat:
image
In this data frame I got the 3rd variable from the integer keyboard, using the divisors key.  I then transformed the data into the 4th variable.  
I would like to have a simple way of evaluating that variable.  We are (I assume) doing such things behind the scenes, because an initial exploration led me to the `pander`, `evaluate` and `lazyeval` packages.  And all these are already installed in R-Instat.
It would be a good addition to the Prepare > Column: Numeric > Transform dialogue and also (of course) in the Calculator.
It would be good to know the feasibility of this soon, even if there isn't time soon to do the programming.  I assume it can be done by someone in Innodems, but @lilyclements might be quicker in working out whether and how, it should be done.

(Currently I use the third variable, a) subset from 3 to -2 to get rid of c(   ), b) then split to put it into multiple columns, c) and then make them numeric, and d) then do a row summary.)

### PR DESCRIPTION
**Pull request by derekagorhom**
Fixes #7886 

This PR is ready for review. 

@rdstern with the exception of the `missing` and `nth` keys, all other keys seems to work fine, also for the maths bloc for the keyboard, i already added the `+,-,*,/ and ^` keys.

### CODE REVIEW

**Review by lloyddewit — DISMISSED** [review]

**Comment by lloyddewit** [comment]
@berylwaswa Do you feel comfortable testing/approving this? 
(please also test all the new tooltips)
thanks

**Comment by berylwaswa** [comment]
@lloyddewit yes I can take a look.

**Comment by lloyddewit** [comment]
> @rdstern @lloyddewit the sumd, missing and nth keys are the only ones not working all other keys seem to work fine.

@berylwaswa Thank you for testing. According to the code changes, it looks like there was an attempt to implement sumd. Should @derekagorhom implement this before we merge?
Should @derekagorhom also implement the `missing` and `nth` keys before we merge or can these be implemented in a separate PR?
thanks

**Comment by berylwaswa** [comment]
@lloyddewit if @derekagorhom is able to implement without the help of @rdstern then he can do that before we merge, if not they can be implemented in a separate PR. 
Thanks.

**Comment by derekagorhom** [comment]
@lloyddewit , @berylwaswa, i will try and solve the issue with the three keys not working so this PR can be left open for now.

**Comment by derekagorhom** [comment]
For the last three commits, i was able to fix the issue with `missing` and `nth keys` not working. The `missing key` should work fine now, for the` nth key` i made sure to add space within the function to enable the user type with number they wanted to display eg.  **sapply( `V1`  , function(x) {dplyr::nth(x, `2` )});** 
Here V1 represents the list and 2 represent the second value in that list (`V1`), (this i hope makes more sense). 
nth key
For the `sumd key` i want to believe roger intended for it to be used on list that have only one value in them but i maybe wrong. 
I also added a new set of keys to the Math bloc of this keyboard, they can be reviewed also. 
thanks

**Comment by derekagorhom** [comment]
@berylwaswa i have made the changes to the code, could you review this

**Comment by berylwaswa** [comment]
@lloyddewit @derekagorhom The missing key and the nth key work fine now.
For the sumd key works only when the list has one value.
The keys added to the math bloc all work fine too.

**Review by berylwaswa — DISMISSED** [review]
These work fine.

**Comment by lloyddewit** [comment]
> For the sumd key works only when the list has one value.

Is this acceptable? or should this still be fixed (in this PR or another PR)?
thanks

**Comment by lloyddewit** [comment]
@berylwaswa @rdstern I just had a quick interaction with @derekagorhom . We agreed that this PR can wait for @rdstern 's testing and feedback.

**Comment by rdstern** [comment]
Looking forward to seeing that.  I am having trouble testing branches on my machine just now.  I can see the current merged version.  So the test will have to wait till I am back on Saturday.

**Review by rdstern — CHANGES_REQUESTED** [review]
@derekagorhom and also @berylwaswa this is looking exciting.  But I am not sure you have made enough effort to understand the different use of sapply and lapply, nor when a function is needed, or not?
So please check the code you have provided, in the keys carefully as follows:
1. lapply produces a list variable, while sapply does not.  So you need lapply for the range (it gives 2 numbers - min and max) and also for the mode(s) , because the mode may not be unique.  You don't need it for those that give a single summary, e.g. mean.
2. You need the function only when it has not been done for you.  So I could not find an ssq function in R - so ssq uses a function.  But there is var and sd, etc, so they can just be given in the same way as you gave min and max on the top row.

So @derekagorhom please examine all the keys now and see what you should do to change them.  Possibly discuss with @berylwaswa too.  This is not so easy, but rewarding when you understand.

As an example I used the factorize2 key on the numbers 1 to 100 in a new dataframe, but getting the factors of x1+100, so for 101, etc.  The example below also shows how I simplified the var key code.

image

**Comment by rdstern** [comment]
The maths section is looking good.
a) Change the round2 code into the round1 key and call it `round`. tooltip "round the numbers in each list. Default is to 3 decimals)
b) Move the code behind round1 to the round2 key and call the key percent2.  Change the 3 to 2 in the code so it rounds percentages to 2 decimals.  Tooltip `percent of each value in the list, rounded to a default of 2 decimals.`  
c) Add the tooltip to the divide key:  `divide each value in the list by a given value (default 10)`
d) Change the square key to say `power (^)`  change the default to raise each number in the list to a given power (default 2, to square the numbers)
d)  Add the symbol to the other keys on the top row sum (+), subtract (-), times (*) divide (/).
e) Change the sqrt function to the simpler lapply(   ,sqrt) tooltip square root each number in the list.  (Change sqrt to abs, log, exp, sin, etc for other functions)
f) Change the word List (above to first key to Summary and move it closer to the first block.  Move it all upwards a bit and put the Help key at the top, moving the Maths block up a bit too.  This to make space for a third block of keys later, called Transform - though we will merge these first.

**Comment by derekagorhom** [comment]
@rdstern i have made the changes but i am still a bit confused on the use of functions for r  for `missing` and `nth`
I have tried to make them simpler like the other r functions but it always returns an error for me.
Below are the links i looked on to make the missing and nth expressions 
[for sum of missing values](https://discuss.analyticsvidhya.com/t/how-to-count-the-missing-value-in-r/2949/5). 
[for the nth value](https://rdrr.io/cran/dplyr/man/nth.html)

**Review by rdstern — CHANGES_REQUESTED** [review]
@derekagorhom it is looking even better now
a) the round doesn't need to be a function, because it is itself!.  So change it to sapply(  ,round,3)
b) the sqrt can be sapply, rather than lapply.
c) for nth, have sapply(  ,dplyr::nth, 2).
d) For first make the tooltip simply `the first value in each list`, for last it is `the last value in each list`, and for nth it is `the nth value in each list, (default = 2, the second value)`
e) In the miss key the code is simply` ,na.rm=TRUE`  Include the comma at the start.  This is not used on its own.  It is to be added to other functions, such as the sum.  The tooltip is `add the option to omit missing values in the summaries`

**Comment by derekagorhom** [comment]
@rdstern i have made the code changes to the `nth, round, miss` and `sqrt` keys and i also added the Transform bloc on the list keyboard.

**Review by rdstern — COMMENTED** [review]
@derekagorhom this has been excellent work.  I have tested lots and all the new keys seem to work fine.

I just suggest a few minor improvements to the tooltips and then it would be great to merge.

a) For the anydup key change row to observation
b) for the quantile key change it to `default is 20%, 50% (median) and 80% of each list`
c) for the fivenum add `gives the same summaries as the boxplot, so minimum, lower and upper hinges (roughly the quartiles), median, and maximum`.  
d) for sumd add gives the sum of the digit sums (e.g. for 34, 27 = 7 + 9 = 16)
e) for round replace by round the numbers in each list, with default of 3 decimals (change round to signif for significant figures)

**Comment by derekagorhom** [comment]
@rdstern I have added the tooltips for the keys on the keyboard.

**Review by rdstern — DISMISSED** [review]
@derekagorhom many thanks for working so well on this new keyboard.  Over to you @lloyddewit

File: instat/ucrCalculator.vb [hunk 1]
```diff
@@ -321,6 +321,54 @@ Public Class ucrCalculator
         ttCalculator.SetToolTip(cmdRan_gamma, "Random sample with shape=1 is from the exponential distribution with mean 2. Keep scale (>0) = 2 and change shape (>0) to 0.5 for chi-square distribution with 1d.f. or to 5, for 10 d.f. Keep scale = 2 for chi-square.")
         ttCalculator.SetToolTip(cmdRan_beta, "This special case is the uniform distribution between 0 and 1. Change shape1 > 0 and shape2 > 0 for different beta distributions.")
         ttCalculator.SetToolTip(cmdRan_sample, "Random sample with replacement, from a given variable. Change replace to FALSE for a random permutation.")
+
+        ttCalculator.SetToolTip(cmdListLength, "number of observations: For example length(c(1,2,3,4,NA)) = 5 ")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:325]: These tooltips seem to be duplicates of the tooltips starting on line 269.
Declare a constant string for each message and then assign the constant string to the tool tips?

File: instat/ucrCalculator.vb [hunk 2]
```diff
@@ -321,6 +321,54 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListSum, "the sum or total: So sum(c(1,2,3,4,10)) = 20 ")
+        ttCalculator.SetToolTip(cmdListMin, "the smallest value: So min(c(4,3,10,1,2)) = 1 ")
+        ttCalculator.SetToolTip(cmdListMax, "the largest value: So max(c(4,3,10,1,2)) = 10 ")
+        ttCalculator.SetToolTip(cmdListRange, "gives both the min and the max. Use max - min to give the value of the range")
+        ttCalculator.SetToolTip(cmdListMissing, "add the option to omit missing values in the summaries")
+        ttCalculator.SetToolTip(cmdListMean, "the average, usually sum/length. So mean(c(1,2,3,4,10)) = 4")
+        ttCalculator.SetToolTip(cmdListMedian, "the value halfway up the values in order. So median(c(1,2,3,4,10) = 3")
+        ttCalculator.SetToolTip(cmdListMode, "the most popular value. So mode(c(10,2,2,3,3) = 2 and 3")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:333]:
```suggestion
        ttCalculator.SetToolTip(cmdListMode, "the most popular value(s). So mode(c(10,2,2,3,3) = 2 and 3")
```

File: instat/ucrCalculator.vb [hunk 3]
```diff
@@ -321,6 +321,54 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListSubtract, "subtract a value (default 10) from each number in the list.")
+        ttCalculator.SetToolTip(cmdListMultiply, "multiply each number in the list by a given value (default 10).")
+        ttCalculator.SetToolTip(cmdDivide, "divide each value in the list by a given value (default 10).")
+        ttCalculator.SetToolTip(cmdListAnyDup, "the observation number of the first duplicated value, or 0 if no duplicates. So anydup(c(1,2,3,3,10) = 4.")
+        ttCalculator.SetToolTip(cmdListSquare, "raise each number in the list to a given power (default 2, to square the numbers).")
+        ttCalculator.SetToolTip(cmdListFirst, "the first value in each list.")
+        ttCalculator.SetToolTip(cmdListLast, "the last value in each list.")
+        ttCalculator.SetToolTip(cmdListNth, "the nth value in each list, (default = 2, the second value).")
+        ttCalculator.SetToolTip(cmdListQuantile, "default is 20%, 50% (median) and 80% of each list.")
+        ttCalculator.SetToolTip(cmdListRound, "replace by round the numbers in each list, with default of 3 decimals (change round to signif for significant figures).")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:351]: this message doesn't make sense to me, rephrase?

File: instat/ucrCalculator.vb [hunk 4]
```diff
@@ -321,6 +321,54 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimals.")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:352]:
```suggestion
        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimals places.")
```

File: instat/ucrCalculator.vb [hunk 5]
```diff
@@ -321,6 +321,54 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListRev, "reverse the data in each list.")
+        ttCalculator.SetToolTip(cmdListOmit, "drop NA values.")
+        ttCalculator.SetToolTip(cmdListIfelse, "replace the condition and the TRUE, FALSE values with those of your choice.")
+        ttCalculator.SetToolTip(cmdListMod, "gives the remainder from division by 10, or a number you choose.  For example 28 %%8 = 4.")
+        ttCalculator.SetToolTip(cmdListDiv, "gives the value after integer division by 10, or a number of your choice.  For example 28 %/% 8 = 3.")
+        ttCalculator.SetToolTip(cmdListOperator, "gives TRUE or FALSE depending on the condition.  (Use >=, <, <=, ==, != as well).")
+        ttCalculator.SetToolTip(cmdListDuplicated, "gives TRUE for any duplicates and FALSE otherwise.")
+        ttCalculator.SetToolTip(cmdListFivenum, "gives the same summaries as the boxplot, so minimum, lower and upper hinges (roughly the quartiles), median, and maximum.")
+        ttCalculator.SetToolTip(cmdListSumd, "gives the sum of the digit sums (e.g. for 34, 27 = 7 + 9 = 16).")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:370]: what does 'for 34' mean?

**Review by lloyddewit — COMMENTED** [review]
@derekagorhom thanks, looks good, just a few small comments

**Comment by derekagorhom** [comment]
@lloyddewit i made changes to the code for the keyboard tooltip and they work fine

**Review by rdstern — CHANGES_REQUESTED** [review]
Just a minor change in the sumd tooltip.  In the brackets it is (e.g. for 34, 27 = 7 + 9 = 16)

File: instat/ucrCalculator.vb [hunk 6]
```diff
@@ -5018,4 +5142,203 @@ Public Class ucrCalculator
     Private Sub cmdDecimals_Click(sender As Object, e As EventArgs) Handles cmdDecimals.Click
         ucrReceiverForCalculation.AddToReceiverAtCursorPosition("sapply(X = , FUN = function(v) {sapply(X = v,FUN = function(w) eval(parse(text=w)))})", 75)
     End Sub
+    Private Sub cmdListLength_Click(sender As Object, e As EventArgs) Handles cmdListLength.Click
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/ucrCalculator.vb:5177]: is this resolved?

**Comment by lloyddewit** [comment]
> @lloyddewit i made changes to the code for the keyboard tooltip and they work fine

Thank you for adding the constant strings.
I made some other comments above also. They were marked as resolved but I checked an they still need resolving.
If you can make the requested changes, then I can approve.
thanks

File: instat/ucrCalculator.vb (hunk 5 above)
Comment (derekagorhom) [reply instat/ucrCalculator.vb:402]: https://github.com/africanmathsinitiative/R-Instat/pull/8102#pullrequestreview-1299824192

this is how the tooltip for that key was made, you can see from the above comment

File: instat/ucrCalculator.vb (hunk 3 above)
Comment (derekagorhom) [reply instat/ucrCalculator.vb]: I changed it to 
"replaces by round, the numbers in each list, with default of 3 decimals (change round to signif for significant figures"
yesterday

Comment (lloyddewit) [reply instat/ucrCalculator.vb]: I still find the message a bit confusing. Is something like thic clearer?

"applies the round function to each value in each list, with a default of 3 decimal places (change round to signif for significant figures)."

**Comment by lloyddewit** [comment]
> Just a minor change in the sumd tooltip. In the brackets it is (e.g. for 34, 27 = 7 + 9 = 16)

@rdstern I'm still a bit confused about 'for 34', what does this mean?

File: instat/ucrCalculator.vb (hunk 5 above)
Comment (lloyddewit) [reply instat/ucrCalculator.vb:402]: now I understand, my mistake, sorry

File: instat/ucrCalculator.vb [hunk 7]
```diff
@@ -321,6 +367,40 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListMultiply, "multiply each number in the list by a given value (default 10).")
+        ttCalculator.SetToolTip(cmdDivide, "divide each value in the list by a given value (default 10).")
+        ttCalculator.SetToolTip(cmdListAnyDup, "the observation number of the first duplicated value, or 0 if no duplicates. So anydup(c(1,2,3,3,10) = 4.")
+        ttCalculator.SetToolTip(cmdListSquare, "raise each number in the list to a given power (default 2, to square the numbers).")
+        ttCalculator.SetToolTip(cmdListFirst, "the first value in each list.")
+        ttCalculator.SetToolTip(cmdListLast, "the last value in each list.")
+        ttCalculator.SetToolTip(cmdListNth, "the nth value in each list, (default = 2, the second value).")
+        ttCalculator.SetToolTip(cmdListQuantile, "default is 20%, 50% (median) and 80% of each list.")
+        ttCalculator.SetToolTip(cmdListRound, "replaces by round, the numbers in each list, with default of 3 decimals (change round to signif for significant figures).")
+        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimals places.")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:384]:
```suggestion
        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimal places.")
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@derekagorhom I think there are 3 comments that still need your action. Please do not set these comments to 'resolved'. Please let me know when you're done and I'll review.
Thanks

**Comment by rdstern** [comment]
@lloyddewit in "Just a minor change in the sumd tooltip. In the brackets it is (e.g. for 34, 27 = 7 + 9 = 16)"
You write 

" I'm still a bit confused about 'for 34', what does this mean?"

If so, then it must need rewording.  I mean if the list contains the 2 numbers 34 and 27, then the sum of the digits (individually) is 3 + 4 = 7 and 2 + 7 = 9, so sumd will give 7 + 9 = 16.

**Comment by derekagorhom** [comment]
so roger can i reword it to 
` if the list contains the 2 numbers 34 and 27, then the sum of the digits (individually) is 3 + 4 = 7 and 2 + 7 = 9, so sumd will give 7 + 9 = 16.`

**Comment by lloyddewit** [comment]
> so roger can i reword it to ` if the list contains the 2 numbers 34 and 27, then the sum of the digits (individually) is 3 + 4 = 7 and 2 + 7 = 9, so sumd will give 7 + 9 = 16.`

@rdstern for me that's perfect. Actually I was being a bit thick. After I wrote the comment, I suddenly understood what it meant and deleted my comment - but you were too fast and had already responded!

**Comment by rdstern** [comment]
@lloyddewit I wondered why I couldn't find your comment in the pull request - but I still had the evidence in the e-mail.
@derekagorhom it is your choice.  I am keen that you understand the use of each key - and the tooltip is designed to help.  So I would like the most concise version that is really clear to you.

File: instat/ucrCalculator.vb [hunk 8]
```diff
@@ -321,6 +367,40 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdDivide, "divide each value in the list by a given value (default 10).")
+        ttCalculator.SetToolTip(cmdListAnyDup, "the observation number of the first duplicated value, or 0 if no duplicates. So anydup(c(1,2,3,3,10) = 4.")
+        ttCalculator.SetToolTip(cmdListSquare, "raise each number in the list to a given power (default 2, to square the numbers).")
+        ttCalculator.SetToolTip(cmdListFirst, "the first value in each list.")
+        ttCalculator.SetToolTip(cmdListLast, "the last value in each list.")
+        ttCalculator.SetToolTip(cmdListNth, "the nth value in each list, (default = 2, the second value).")
+        ttCalculator.SetToolTip(cmdListQuantile, "default is 20%, 50% (median) and 80% of each list.")
+        ttCalculator.SetToolTip(cmdListRound, "applies the round function to each value in each list, with a default of 3 decimal places (change round to signif for significant figures).")
+        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimals.")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:384]:
```suggestion
        ttCalculator.SetToolTip(cmdListPercent2, "percent of each value in the list, rounded to a default of 2 decimal places.")
```

File: instat/ucrCalculator.vb [hunk 9]
```diff
@@ -321,6 +367,40 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdListRev, "reverse the data in each list.")
+        ttCalculator.SetToolTip(cmdListOmit, "drop NA values.")
+        ttCalculator.SetToolTip(cmdListIfelse, "replace the condition and the TRUE, FALSE values with those of your choice.")
+        ttCalculator.SetToolTip(cmdListMod, "gives the remainder from division by 10, or a number you choose.  For example 28 %%8 = 4.")
+        ttCalculator.SetToolTip(cmdListDiv, "gives the value after integer division by 10, or a number of your choice.  For example 28 %/% 8 = 3.")
+        ttCalculator.SetToolTip(cmdListOperator, "gives TRUE or FALSE depending on the condition.  (Use >=, <, <=, ==, != as well).")
+        ttCalculator.SetToolTip(cmdListDuplicated, "gives TRUE for any duplicates and FALSE otherwise.")
+        ttCalculator.SetToolTip(cmdListFivenum, "gives the same summaries as the boxplot, so minimum, lower and upper hinges (roughly the quartiles), median, and maximum.")
+        ttCalculator.SetToolTip(cmdListSumd, " if the list contains the 2 numbers 34 and 27, then the sum of the digits (individually) is 3 + 4 = 7 and 2 + 7 = 9, so sumd will give 7 + 9 = 16.")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:402]:
```suggestion
        ttCalculator.SetToolTip(cmdListSumd, "if the list contains the 2 numbers 34 and 27, then the sum of the digits (individually) is 3 + 4 = 7 and 2 + 7 = 9, so sumd will give 7 + 9 = 16.")
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern there are some small changes since your last approval. If you can retest/approve, then we can merge, thanks

**Review by rdstern — APPROVED** [review]
@derekagorhom and @lloyddewit this looks very good.  Many thanks
